File: CPP/ByRound/1968/1968_D_Permutation_Game.cpp
// Problem: CF 1968 D - Permutation Game
// https://codeforces.com/contest/1968/problem/D

/*
D. Permutation Game

Algorithms/Techniques:
- Simulation with cycle detection in permutation
- Prefix sums for efficient score calculation
- Optimization using cycle length to avoid unnecessary iterations when k is large

Time Complexity: O(n) per test case, where n is the length of the permutation.
Space Complexity: O(n) for storing the permutation and array.

The game involves two players moving through a permutation with a fixed number of turns,
and each turn contributes to the score based on the value of the current position.
We simulate both players' moves, taking advantage of cycles in the permutation to
avoid redundant computation when k is large.
*/

#include <algorithm>
#include <cstdio>
#include <vector>
typedef long long ll;

int main() {
  long t;
  scanf("%ld", &t);
  while (t--) {
    ll n, k, pb, ps;
    scanf("%lld %lld %lld %lld", &n, &k, &pb, &ps);
    std::vector<ll> m(n + 1); // permutation array
    for (ll p = 1; p <= n; p++) {
      scanf("%lld", &m[p]);
    }
    std::vector<ll> a(n + 1); // score array
    for (ll p = 1; p <= n; p++) {
      scanf("%lld", &a[p]);
    }
    
    // Bodya's simulation
    ll idxb(pb), csb(0), kb(k), scoreb(0), mxb(0);
    do {
      scoreb = std::max(scoreb, csb + kb * a[idxb]); // update max score with potential future bonus
      csb += a[idxb]; // accumulate current score
      --kb;
      idxb = m[idxb]; // move to next position in permutation
    } while (kb && idxb != pb); // stop if no more moves or back to start

    // Sasha's simulation
    ll idxs(ps), css(0), ks(k), scores(0), mxs(0);
    do {
      scores = std::max(scores, css + ks * a[idxs]); // update max score with potential future bonus
      css += a[idxs]; // accumulate current score
      --ks;
      idxs = m[idxs]; // move to next position in permutation
    } while (ks && idxs != ps); // stop if no more moves or back to start

    if (scoreb == scores) {
      puts("Draw");
    } else {
      puts(scoreb > scores ? "Bodya" : "Sasha");
    }
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/1968/1968_D_Permutation_Game.py
# Problem: CF 1968 D - Permutation Game
# https://codeforces.com/contest/1968/problem/D

"""
Algorithm/Technique: 
- The problem involves simulating a game between two players (Bodya and Sasha) on a permutation with a fixed number of moves (k).
- Each player starts at a position and can either stay or move according to the permutation array.
- The score is accumulative, with each move adding the value at the current position.
- The key insight is to use a simulation with cycle detection to avoid unnecessary computation for large k.
- We track the best possible score in all possible paths up to k turns, leveraging properties of cycles in permutations.

Time Complexity: O(n + k) per test case, where n is the permutation length and k is the number of turns.
Space Complexity: O(n) for storing the permutation, scores, and tracking seen positions.
"""

def solve():
    n, k, pb, ps = map(int, input().split())
    pb -= 1
    ps -= 1
    p = list(map(int, input().split()))
    p = [x - 1 for x in p]  # Convert to 0-based indexing
    score = list(map(int, input().split()))
    mx = max(score)
    d = set([i for i in range(n) if score[i] == mx])  # Set of indices with maximum score

    def get_score(pos, k):
        ans = 0
        now = 0
        seen = set()
        while k:
            # Update max possible score considering remaining turns
            ans = max(ans, now + score[pos] * k)
            seen.add(pos)
            # If not in a cycle (no repeated path) and can move, do so
            if pos not in d and p[pos] not in seen:
                now += score[pos]
                pos = p[pos]
            else:
                # If in a cycle or final step, stop early to optimize
                break
            k -= 1
        return ans

    sb, ss = get_score(pb, k), get_score(ps, k)
    if sb > ss:
        print("Bodya")
    elif sb < ss:
        print("Sasha")
    else:
        print("Draw")


T = int(input())
for _ in range(T):
    solve()


# https://github.com/VaHiX/CodeForces/